drivers: sdhc: Add driver support for xlnx SDHC

AlvinPaul26 opened the pull request:
Add driver support for xlnx SD/EMMC host controller. The driver currently support SD host controller version 3.0 and EMMC host controller version 5.1. This driver functions with the SDHC subsystem to perform operations on device.

Driver support both interrupt and polled mode data transfer. Uses ADMA2 to perform data transfer.

AlvinPaul26 committed:
drivers: sdhc: Add driver support for xlnx SDHC

Add driver support for xlnx SD/EMMC host controller. The driver
currently support SD host controller version 3.0 and EMMC host
controller version 5.1. This driver functions with the SDHC subsystem to
perform operations on device.

Driver support both interrupt and polled mode data transfer. Uses ADMA2
to perform data transfer.

Signed-off-by: Paul Alvin <alvin.paulp@amd.com>

michalsimek reviewed (approved):
I have tested in it on Qemu and already reviewed this code before submission to upstream and no issue from my side.

danieldegrasse commented on drivers/sdhc/xlnx_sdhc.c:
```diff
@@ -0,0 +1,1361 @@
+	}
+
+	/* Get input clock rate */
+	ret = clock_control_get_rate(config->clock_dev, NULL, &config->maxclock);
```
The `maxclock` value should be stored in a data pointer- the `config` pointer may be in ROM on some platforms

AlvinPaul26 replied:
thanks for the review.. will move it to data pointer.

danieldegrasse commented on drivers/sdhc/xlnx_sdhc.c:
```diff
@@ -0,0 +1,1361 @@
+	/**< Maximum input clock supported by HC */
+	uint32_t maxclock;
+	/**< ADMA descriptor table */
+	adma2_descriptor adma2_descrtbl[MAX(1, CONFIG_HOST_ADMA2_DESC_SIZE)];
```
This looks like it is modified at runtime- it should be in a `data` struct- the `config` struct should be used for RO data that can be placed into flash (ROM) on platforms that support it.

AlvinPaul26 replied:
will move all mutable to data pointer.

danieldegrasse commented on drivers/sdhc/xlnx_sdhc.c:
```diff
@@ -0,0 +1,1361 @@
+ */
+static int8_t xlnx_sdhc_cmd_response(const struct device *dev, struct sdhc_command *cmd)
+{
+	struct xlnx_sdhc_config *config = (struct xlnx_sdhc_config *)dev->config;
```
We should always be using `const struct xlnx_sdhc_config`. Essentially the config structure should store read only const data, and the `dev->data` pointer should point to a mutable data structure in RAM

AlvinPaul26 replied:
will update it.

danieldegrasse commented:
Thanks for making the updates here- is there a platform this was built against, or is it only being used downstream?

AlvinPaul26 commented:
> 

It can be build with "boards/amd/versalnet_rpu" platform. The driver is validated with "test/subsys/sd/mmc" and "test/subsys/sd/sdmmc" testcases on hardware. At the time of rising this pull request 'versalnet_rpu' platform support is not merged so I didn't added the nodes into platform dts. Do you want me to add the nodes to the platform dts ?

AlvinPaul26 commented:
> > > 
> > 
> > 
> > It can be build with "boards/amd/versalnet_rpu" platform. The driver is validated with "test/subsys/sd/mmc" and "test/subsys/sd/sdmmc" testcases on hardware. At the time of rising this pull request 'versalnet_rpu' platform support is not merged so I didn't added the nodes into platform dts. Do you want me to add the nodes to the platform dts ?
> 
> Could you point me to this PR? I'd ideally like to see this enabled on some platform in upstream so it has CI coverage- ideally I'd prefer to see the versalnet_rpu platform merge first, and than have this enabled there. If that's not possible, we can look at getting this building on `native_sim` and add a test for that

 versalnet_rpu platform support has been merged. Will enable this driver there and send the next version.

AlvinPaul26 committed:
dts: Add device tree support for SDHC

Add sdhci0(SD) and sdhci1(EMMC) bus node with disk support to DT for
Versal NET.

Signed-off-by: Paul Alvin <alvin.paulp@amd.com>

AlvinPaul26 committed:
boards: amd: Enable SDHC DT nodes

Enable both sdhci0 and sdhci1 DT node. IP is tested on HW for Versal
NET.

Signed-off-by: Paul Alvin <alvin.paulp@amd.com>

AlvinPaul26 committed:
tests: sd: mmc: Add filter for sdhc0 alias

Test expect sdhc0 alias to be defined.

Signed-off-by: Paul Alvin <alvin.paulp@amd.com>

AlvinPaul26 committed:
tests: subsys: sd: Add board overlays to support testcase

Add SD/EMMC overlay files to support testcases for Versal NET. This
overlays are used to select the device to be tested.

Signed-off-by: Paul Alvin <alvin.paulp@amd.com>

danieldegrasse commented on tests/subsys/sd/mmc/boards/versalnet_rpu.overlay:
Can you add alias at the board level? This alias is used in several tests

danieldegrasse commented:
@decsny could you take a look at this?
